nameko/utils.py

```diff
@@ -1,9 +1,108 @@
+import inspect
+import re
 
 import eventlet
 from eventlet.queue import LightQueue
 
 
+REDACTED = "********"
+
+
+def get_redacted_args(entrypoint, *args, **kwargs):
+    """ Utility function for use with entrypoints that are marked with
+    ``sensitive_variables`` -- e.g. :class:`nameko.rpc.Rpc` and
+    :class:`nameko.events.EventHandler`.
+
+    :Parameters:
+        entrypoint : :class:`~nameko.extensions.Entrypoint`
+            The entrypoint that fired.
+        args : tuple
+            Positional arguments for the method call.
+        kwargs : dict
+            Keyword arguments for the method call.
+
+    The entrypoint should have a ``sensitive_variables`` attribute, the value
+    of which is a string or tuple of strings specifying the arguments or
+    partial arguments that should be redacted. To partially redact an argument,
+    the following syntax is used::
+
+        <argument-name>.<dict-key>[<list-index>]
+
+    :Returns:
+        A dictionary as returned by :func:`inspect.getargspec`, but with
+        sensitive arguments or partial arguments redacted.
+
+    .. note::
+
+        This function does not raise if one of the ``sensitive_variables``
+        doesn't match or partially match the calling ``args`` and ``kwargs``.
+        This allows "fuzzier" pattern matching (e.g. redact a field if it is
+        present, and otherwise do nothing).
+
+        To avoid exposing sensitive variables through a typo, it is recommend
+        to test the configuration of each entrypoint with
+        ``sensitive_variables`` individually. For example:
+
+        .. code-block:: python
+
+            class Service(object):
+                @rpc(sensitive_variables="foo.bar")
+                def method(self, foo):
+                    pass
+
+            container = ServiceContainer(Service, {})
+            entrypoint = get_extension(container, Rpc, method_name="method")
+
+            # no redaction
+            foo = "arg"
+            expected_foo = {'foo': "arg"}
+            assert get_redacted_args(entrypoint, foo) == expected
+
+            # 'bar' key redacted
+            foo = {'bar': "secret value", 'baz': "normal value"}
+            expected = {'foo': {'bar': "********", 'baz': "normal value"}}
+            assert get_redacted_args(entrypoint, foo) == expected
+
+    .. seealso::
+
+        The tests for this utility demonstrate its full usage:
+        :class:`test.test_utils.TestGetRedactedArgs`
+
+    """
+    sensitive_variables = entrypoint.sensitive_variables
+    if isinstance(sensitive_variables, basestring):
+        sensitive_variables = (sensitive_variables,)
+
+    method = getattr(entrypoint.container.service_cls, entrypoint.method_name)
+    argspec = inspect.getcallargs(method, None, *args, **kwargs)
+    del argspec['self']
+
+    def redact(data, keys):
+        key = keys[0]
+        if len(keys) == 1:
+            try:
+                data[key] = REDACTED
+            except (KeyError, IndexError):
```

Comment:
you can also get `TypeError` here if the argument doesn't match (e.g. trying to redact an integer thinking it's a dict)